tidb restart failed cause by panic

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sh
# create a tidb cluster
tiup cluster deploy csdb nightly topo.yaml -y
tiup cluster start csdb -R tidb
# use br to load data
tiup br restore db --checksum=false --db=oltp_12x1e7 --pd=http://10.202.0.41:2379/ --storage=gs://oltp-bench-us-east5/dataset/v8.1.0/oltp_12x1e7
# restart tidb cluster
tiup cluster restart csdb  # tidb restart failed cause by panic
```

```log
Error: failed to start: failed to start tidb: failed to start: 10.202.0.38 tidb-4000.service, please check the instance's log(/data/deploy/tidb-4000/log) for more detail.: timed out waiting for port 4000 to be started after 2m0s
```

Related tidb_stderr.log log:

```log
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x48b880b]

goroutine 713 [running]:
[github.com/pingcap/tidb/pkg/statistics/handle.(*Handle).initStatsHistograms4Chunk](http://github.com/pingcap/tidb/pkg/statistics/handle.(*Handle).initStatsHistograms4Chunk)(0xc00179a400, {0x7148938, 0xc001187c28}, {0x712ee10, 0xc000015590}, 0xc003a2bd88, 0x0)
        /workspace/source/tidb/pkg/statistics/handle/bootstrap.go:233 +0x4ab
[github.com/pingcap/tidb/pkg/statistics/handle.(*Handle).initStatsHistogramsByPaging](http://github.com/pingcap/tidb/pkg/statistics/handle.(*Handle).initStatsHistogramsByPaging)(0xc00179a400, {0x7148938, 0xc001187c28}, {0x712ee10, 0xc000015590}, {0xc003468100?, 0xc003a1d6a8?}, 0xf92af2000)
        /workspace/source/tidb/pkg/statistics/handle/bootstrap.go:360 +0x425
[github.com/pingcap/tidb/pkg/statistics/handle.(*Handle).initStatsHistogramsConcurrency.func1](http://github.com/pingcap/tidb/pkg/statistics/handle.(*Handle).initStatsHistogramsConcurrency.func1)({0x0?, 0x0?})
        /workspace/source/tidb/pkg/statistics/handle/bootstrap.go:369 +0x37
[github.com/pingcap/tidb/pkg/statistics/handle/initstats.(*RangeWorker).loadStats(0xc00342b1a0)](http://github.com/pingcap/tidb/pkg/statistics/handle/initstats.(*RangeWorker).loadStats(0xc00342b1a0))
        /workspace/source/tidb/pkg/statistics/handle/initstats/load_stats_page.go:91 +0x70
[github.com/pingcap/tidb/pkg/statistics/handle/initstats.(*RangeWorker).LoadStats.func1()](http://github.com/pingcap/tidb/pkg/statistics/handle/initstats.(*RangeWorker).LoadStats.func1())
        /workspace/source/tidb/pkg/statistics/handle/initstats/load_stats_page.go:84 +0x17
[github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1()](http://github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1())
        /workspace/source/tidb/pkg/util/wait_group_wrapper.go:157 +0x4c
created by [github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run](http://github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run) in goroutine 647
        /workspace/source/tidb/pkg/util/wait_group_wrapper.go:155 +0x73
```


<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

cluster restart success.

### 3. What did you see instead (Required)

```sh
Error: failed to start: failed to start tidb: failed to start: 10.202.0.38 tidb-4000.service, please check the instance's log(/data/deploy/tidb-4000/log) for more detail.: timed out waiting for port 4000 to be started after 2m0s
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```sql
> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                     |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v8.5.0-alpha-373-g177a03c8e5
Edition: Community
Git Commit Hash: 177a03c8e51caf844491d8a93edda29295510e8d
Git Branch: HEAD
UTC Build Time: 2024-12-18 05:08:08
GoVersion: go1.23.4
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

--- a/pkg/statistics/handle/bootstrap.go
+++ b/pkg/statistics/handle/bootstrap.go
@@ -164,6 +164,12 @@ func (*Handle) initStatsHistograms4ChunkLite(cache statstypes.StatsCache, iter *
 }
 
 func (h *Handle) initStatsHistograms4Chunk(is infoschema.InfoSchema, cache statstypes.StatsCache, iter *chunk.Iterator4Chunk, isCacheFull bool) {
+	defer func() {
+		if r := recover(); r != nil {
+			logutil.BgLogger().Error("panic when initStatsHistograms4Chunk", zap.Any("r", r),
+				zap.Stack("stack"))
+		}
+	}()
 	var table *statistics.Table
 	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
 		tblID, statsVer := row.GetInt64(0), row.GetInt64(8)

--- a/pkg/statistics/handle/bootstrap.go
+++ b/pkg/statistics/handle/bootstrap.go
@@ -193,6 +193,7 @@ func (h *Handle) initStatsHistograms4Chunk(is infoschema.InfoSchema, cache stats
 		lastAnalyzePos := row.GetDatum(11, types.NewFieldType(mysql.TypeBlob))
 		tbl, ok := h.TableInfoByID(is, table.PhysicalID)
 		if !ok {
+			// this table has been dropped. but stats meta still exists and wait for being deleted.
 			logutil.BgLogger().Warn("cannot find this table when to init stats", zap.Int64("tableID", table.PhysicalID))
 			continue
 		}

--- a/pkg/statistics/handle/bootstrap.go
+++ b/pkg/statistics/handle/bootstrap.go
@@ -185,7 +185,11 @@ func (h *Handle) initStatsHistograms4Chunk(is infoschema.InfoSchema, cache stats
 		}
 		id, ndv, nullCount, version, totColSize := row.GetInt64(2), row.GetInt64(3), row.GetInt64(5), row.GetUint64(4), row.GetInt64(7)
 		lastAnalyzePos := row.GetDatum(11, types.NewFieldType(mysql.TypeBlob))
-		tbl, _ := h.TableInfoByID(is, table.PhysicalID)
+		tbl, ok := h.TableInfoByID(is, table.PhysicalID)
+		if !ok {
+			logutil.BgLogger().Warn("cannot find this table when to init stats", zap.Int64("tableID", table.PhysicalID))
+			continue
+		}
 		if row.GetInt64(1) > 0 {
 			var idxInfo *model.IndexInfo
 			for _, idx := range tbl.Meta().Indices {

--- a/pkg/statistics/handle/handletest/initstats/BUILD.bazel
+++ b/pkg/statistics/handle/handletest/initstats/BUILD.bazel
@@ -8,6 +8,7 @@ go_test(
         "main_test.go",
     ],
     flaky = True,
+    shard_count = 4,
     deps = [
         "//pkg/config",
         "//pkg/parser/model",

--- a/pkg/statistics/handle/handletest/initstats/load_stats_test.go
+++ b/pkg/statistics/handle/handletest/initstats/load_stats_test.go
@@ -27,7 +27,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestConcurrentlyInitStats(t *testing.T) {
+func TestConcurrentlyInitStatsWithMemoryLimit(t *testing.T) {
 	restore := config.RestoreFunc()
 	defer restore()
 	config.UpdateGlobal(func(conf *config.Config) {

--- a/pkg/statistics/handle/handletest/initstats/load_stats_test.go
+++ b/pkg/statistics/handle/handletest/initstats/load_stats_test.go
@@ -27,7 +27,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestConcurrentlyInitStatsWithMemoryLimit(t *testing.T) {
+func TestConcurrentlyInitStats(t *testing.T) {
 	restore := config.RestoreFunc()
 	defer restore()
 	config.UpdateGlobal(func(conf *config.Config) {
@@ -103,7 +103,27 @@ func testConcurrentlyInitStats(t *testing.T) {
 	require.Equal(t, int64(126), handle.GetMaxTidRecordForTest())
 }
 
-func TestIssue58371(t *testing.T) {
+func TestDropTableBeforeConcurrentlyInitStats(t *testing.T) {
+	restore := config.RestoreFunc()
+	defer restore()
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.Performance.LiteInitStats = false
+		conf.Performance.ConcurrentlyInitStats = true
+	})
+	testDropTableBeforeInitStats(t)
+}
+
+func TestDropTableBeforeNonLiteInitStats(t *testing.T) {
+	restore := config.RestoreFunc()
+	defer restore()
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.Performance.LiteInitStats = false
+		conf.Performance.ConcurrentlyInitStats = false
+	})
+	testDropTableBeforeInitStats(t)
+}
+
+func testDropTableBeforeInitStats(t *testing.T) {
 	store, dom := testkit.CreateMockStoreAndDomain(t)
 	tk := testkit.NewTestKit(t, store)
 	tk.MustExec("use test;")

--- a/pkg/statistics/handle/handletest/initstats/load_stats_test.go
+++ b/pkg/statistics/handle/handletest/initstats/load_stats_test.go
@@ -102,3 +102,20 @@ func testConcurrentlyInitStats(t *testing.T) {
 	}
 	require.Equal(t, int64(126), handle.GetMaxTidRecordForTest())
 }
+
+func TestIssue58371(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test;")
+	tk.MustExec("create table t( id int, a int, b int, index idx(id, a));")
+	tk.MustExec("insert into t values (1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4), (5, 5, 5);")
+	tk.MustExec("insert into t select * from t where id<>2;")
+	tk.MustExec("insert into t select * from t where id<>2;")
+	tk.MustExec("insert into t select * from t where id<>2;")
+	tk.MustExec("insert into t select * from t where id<>2;")
+	tk.MustExec("analyze table t all columns;")
+	tk.MustExec("drop table t")
+	h := dom.StatsHandle()
+	is := dom.InfoSchema()
+	require.NoError(t, h.InitStats(context.Background(), is))
+}

